scripts/challenge_week.py

```diff
@@ -0,0 +1,260 @@
+
+        Returns:
+            df_train: dataframe containing all weather data, pv_power data, and demand data
+                      given for the task week. There will be missing weather data at this point,
+                      becuase the weather data is hourly.
+            df_taskweek: dataframe containing all the wather data that can be used for
+                         prediction of the wekk of interest.
+
+        '''
+        # Check 'datetime' columns are equal in demand and pv dfs, this must be true
+        assert self.df_demand['datetime'].equals(self.df_pv['datetime']), "Demand and PV train sets have different timestamps"
+
+        # Merge the demand and pv_power dataset together into one dataframe
+        self.df_train = self.df_pv.merge(right=self.df_demand,
+                                         on='datetime',
+                                         how='left')
+
+        # Create a weather sets that matches the date range of pv_demand set and submit set
+        # Set DatetimeIndex
+        self.df_weather.set_index('datetime',
+                                  inplace=True)
+
+        # Create weather dataframe with matching timestamps with training/testing data
+        start_date = self.df_train.loc[0, 'datetime']
+        end_date = self.df_train.loc[len(self.df_train)-1, 'datetime']
+        df_weather_train = self.df_weather.loc[start_date:end_date]
+
+        # Create weather dataframe with matching timestamps with task week submission data
+        start_date = self.df_submit.loc[0, 'datetime']
+        end_date = self.df_submit.loc[len(self.df_submit)-1, 'datetime']
+        df_weather_taskweek = self.df_weather.loc[start_date:end_date]
+
+        # Join pv_demand set with weather_set0 set to complete the model dataframe
+        self.df_train = self.df_train.merge(right=df_weather_train,
+                                            on='datetime',
+                                            how='outer')
+        self.df_taskweek = self.df_submit.merge(right=df_weather_taskweek,
+                                                  on='datetime',
+                                                  how='outer')
+
+        return self.df_train, self.df_taskweek
+
+    def impute_weather(self, lst_cols, df):
+        '''Imputes the missing data for the weather paramters using the mean of the
+        timesteps before and after.
+
+        Args:
+            lst_cols(list): list of weather columns to impute
+            df: dataframe to which to pply the transformation
+        Returns:
+            df: transformed df
+        '''
+        for c in range(0, len(lst_cols)):
+            for i in range(0, len(df)-1):
+                if pd.isnull(df[lst_cols[c]].iloc[i]):
+                    df.loc[i, lst_cols[c]] = mean([df.loc[i-1, lst_cols[c]], df.loc[i+1, lst_cols[c]]])
+            df[lst_cols[c]].iloc[len(df)-1] = df[lst_cols[c]].iloc[len(df)-2]
```

Comment:
Sorry, I'm lost as to what this line is doing.  Why rewrite the last value of each column with the one before it (if I am reading this correctly)?

Reply:
With the method I am using, if the last value in the column was NaN, it wouldn't be able to find the mean using the values immediately before and after, so I just assigned it the same value as the value immediately before it. Shouldnt be a huge deal considering the weather parameters usually wouldnt change too much in 30 minutes, but hopefully the Bfill and Ffill method could help with some of the short comings of this method.

Reply:
Ok, that makes sense, but isn't it outside the block that checks if that entry is NaN?